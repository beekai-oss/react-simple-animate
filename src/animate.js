// @flow
import React, { createElement } from 'react';

export const defaultState = {
  animationWillEnd: false,
  animationWillStart: false,
  animationWillComplete: false,
  played: false,
};

type Style = { [string]: string | number };

type Props = {
  startAnimation: boolean,
  children?: any,
  startStyle?: Style,
  endStyle: Style,
  onCompleteStyle?: Style,
  durationSeconds: number,
  delaySeconds?: number,
  reverseDelaySeconds?: number,
  easeType: string,
  forceUpdate?: boolean,
  tag?: ?string,
  onComplete?: () => mixed,
  className?: string,
  transition?: string,
};

type State = {
  animationWillEnd: boolean,
  animationWillStart: boolean,
  animationWillComplete: boolean,
  played: boolean,
};

export default class Animate extends React.Component<Props, State> {
  static defaultProps = {
    durationSeconds: 0.3,
    delaySeconds: 0,
    easeType: 'linear',
    reverseDelaySeconds: 0,
    tag: 'div',
  };

  state = defaultState;

  animationTimeout = null;

  animationCompleteTimeout = null;

  setAnimationDelay = (
    condition: boolean,
    durationSeconds: number,
    phase: 'play' | 'reverse',
  ): void => {
    if (!condition) return;
    clearTimeout(this.animationTimeout);
    this.animationTimeout = setTimeout(() => {
      this.setState({
        ...(phase === 'play' ? { animationWillEnd: true } : null),
        ...(phase === 'reverse' ? { animationWillStart: true } : null),
      });
    }, parseFloat(durationSeconds) * 1000);
  };

  setAnimationDelayAndOnComplete(props: Props, isReverse: boolean = false) {
    const {
      delaySeconds,
      startAnimation,
      onCompleteStyle,
      durationSeconds,
      onComplete,
      reverseDelaySeconds,
    } = props;

    const delayTotalSeconds =
      parseFloat(delaySeconds) + parseFloat(durationSeconds);

    // delay animation
    this.setAnimationDelay(
      !!delaySeconds && startAnimation,
      delaySeconds || 0,
      'play',
    );
    // reverse animation
    this.setAnimationDelay(
      !!reverseDelaySeconds && isReverse && !startAnimation,
      reverseDelaySeconds || 0,
      'reverse',
    );

    if ((!onComplete && !onCompleteStyle) || !startAnimation) return;

    this.animationCompleteTimeout = setTimeout(() => {
      if (onComplete) onComplete();
      if (onCompleteStyle) {
        this.setState({
          animationWillComplete: true,
        });
      }
    }, delayTotalSeconds * 1000);
  }

  componentWillReceiveProps(nextProps: Props) {
    const { startAnimation } = nextProps;
    const isAnimationStatusChanged =
      startAnimation !== this.props.startAnimation;

    if (isAnimationStatusChanged) {
      this.setState({
        animationWillEnd: false,
        animationWillStart: false,
        animationWillComplete: false,
        played: isAnimationStatusChanged && startAnimation,
      });
    }

    this.setAnimationDelayAndOnComplete(
      nextProps,
      isAnimationStatusChanged && !startAnimation,
    );
  }

  componentDidMount() {
    this.setAnimationDelayAndOnComplete(this.props);
  }

  shouldComponentUpdate(nextProps: Props, nextState: State) {
    // only situation that should trigger a re-render
    return (
      JSON.stringify(nextProps.startStyle) !==
        JSON.stringify(this.props.startStyle) ||
      JSON.stringify(nextProps.endStyle) !==
        JSON.stringify(this.props.endStyle) ||
      nextProps.startAnimation !== this.props.startAnimation ||
      nextProps.children !== this.props.children ||
      nextState.animationWillEnd !== this.state.animationWillEnd ||
      nextState.animationWillStart !== this.state.animationWillStart ||
      nextState.animationWillComplete !== this.state.animationWillComplete ||
      !!nextProps.forceUpdate
    );
  }

  componentWillUnmount() {
    clearTimeout(this.animationTimeout);
    clearTimeout(this.animationCompleteTimeout);
    this.animationTimeout = null;
    this.animationCompleteTimeout = null;
  }

  render() {
    const {
      animationWillEnd,
      animationWillStart,
      animationWillComplete,
      played,
    } = this.state;
    const {
      children,
      startAnimation,
      startStyle,
      endStyle,
      onCompleteStyle,
      durationSeconds,
      reverseDelaySeconds,
      delaySeconds,
      easeType,
      className,
      transition: transitionValue,
      tag,
    } = this.props;
    let style = startStyle;
    let transition = transitionValue || `${durationSeconds}s all ${easeType}`;

    if (!played && reverseDelaySeconds && !startAnimation) {
      style = animationWillStart ? startStyle : endStyle;
    } else if (
      animationWillComplete ||
      animationWillEnd ||
      (startAnimation && !delaySeconds)
    ) {
      if (onCompleteStyle && animationWillComplete) {
        style = onCompleteStyle;
        transition = null;
      } else {
        style = endStyle;
      }
    }

    const propsToChild = {
      className,
      style: {
        ...{
          ...style,
          transition,
        },
      },
    };

    return createElement(tag || 'div', propsToChild, children);
  }
}
