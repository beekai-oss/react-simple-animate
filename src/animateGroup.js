// @flow
import React, { type Element } from 'react';
import type { AnimationType, AnimationStateType } from './animate';
import calculateTotalDuration from './utils/calculateTotalDuration';

type Sequence = AnimationType & {
  sequenceId?: string | number,
  sequenceIndex?: number,
};

type Sequences = Array<Sequence>;

type Props = {
  play: boolean,
  sequences?: Sequences,
  reverseSequences?: Sequences,
  children: Element<*>,
};

type State = {
  animationStates?: AnimationStateType,
};

// Flowtype to be added later
// <{
//   animationStates?: AnimationStateType,
//   register: any => void,
// }>
// $FlowIgnoreLine
export const AnimateContext = React.createContext({
  animationStates: {},
  register: () => {},
});

export default class AnimateGroup extends React.PureComponent<Props, State> {
  static displayName = 'AnimateGroup';

  static defaultProps = {
    sequences: [],
    reverseSequences: [],
  };

  state = {
    animationStates: {},
  };

  timers: { [string | number]: TimeoutID } = {};

  animations: { [string | number]: AnimationType } = {};

  componentDidMount() {
    this.props.play && this.calculateSequences();
  }

  componentDidUpdate(prevProps: Props) {
    if (this.props.play !== prevProps.play) this.calculateSequences();
  }

  componentWillUnmount() {
    Object.values(this.timers).forEach((timer: any) => clearTimeout(timer));
  }

  setupAnimationTimers = ({
    totalDuration,
    id,
    restAttributes,
    play,
  }: {
    totalDuration: number,
    id: number | string,
    restAttributes: AnimationType,
    play: boolean,
  }) => {
    this.timers[id] = setTimeout(() => {
      this.setState(previousState => {
        const stateCopy = { ...previousState.animationStates };

        if (!stateCopy[id]) stateCopy[id] = {};
        Object.entries(restAttributes).forEach(([key, value]) => {
          stateCopy[id][key] = value;
        });
        stateCopy[id].play = play;

        return {
          animationStates: stateCopy,
        };
      });
    }, totalDuration);
  };

  calculateSequences = () => {
    const { sequences, reverseSequences, play } = this.props;
    const sequencesToAnimate: any =
      Array.isArray(sequences) && sequences.length ? sequences : Object.values(this.animations);
    const reverseSequencesToAnimate: any =
      Array.isArray(reverseSequences) && reverseSequences.length ? reverseSequences : [...sequencesToAnimate].reverse();

    return (play ? sequencesToAnimate : reverseSequencesToAnimate).reduce((previous, current, currentIndex) => {
      const { sequenceId, sequenceIndex, ...restAttributes } = current;
      const id = sequenceId === undefined && sequenceIndex === undefined ? currentIndex : sequenceId || sequenceIndex;
      const totalDuration =
        previous +
        calculateTotalDuration({
          ...this.animations[id],
          play,
          restAttributes,
        });

      this.setupAnimationTimers({
        id,
        totalDuration: currentIndex === 0 ? 0 : previous,
        restAttributes,
        play,
      });
      return totalDuration;
    }, 0);
  };

  register = (props: Sequence) => {
    const { sequenceIndex, sequenceId } = props;
    const id = sequenceId || sequenceIndex;
    if (id === undefined || (sequenceIndex && sequenceIndex < 0) || (sequenceId && sequenceId === '')) return;

    this.animations[id] = {
      ...props,
    };
  };

  render() {
    return (
      <AnimateContext.Provider value={{ animationStates: this.state.animationStates, register: this.register }}>
        {this.props.children}
      </AnimateContext.Provider>
    );
  }
}
