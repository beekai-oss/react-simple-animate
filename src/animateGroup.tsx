import * as React from 'react';
import { Sequences, AnimationProps, AnimateKeyframesProps } from './types';
import calculateTotalDuration from './utils/calculateTotalDuration';
import getSequenceId from './utils/getSequenceId';

const { useState, useRef, useEffect } = React;

export interface Props {
  play: boolean;
  sequences?: Sequences;
  children?: any;
}

export const AnimateContext = React.createContext({
  animationStates: {},
  register: (data: AnimationProps | AnimateKeyframesProps): void => {},
});

export default function AnimateGroup(props: Props) {
  const { play, sequences = [], children } = props;
  const [animationStates, setAnimationStates] = useState();
  const animationsRef = useRef<{ [key: string]: AnimationProps | AnimateKeyframesProps }>({});

  const register = (data: AnimationProps | AnimateKeyframesProps) => {
    const { sequenceIndex, sequenceId } = data;
    if (sequenceId === undefined && sequenceIndex === undefined) return;

    animationsRef.current[getSequenceId(sequenceIndex, sequenceId)] = data;
  };

  useEffect(
    (): void => {
      const sequencesToAnimate =
        Array.isArray(sequences) && sequences.length ? sequences : Object.values(animationsRef.current);
      const localAnimationState = {};

      (play ? sequencesToAnimate : [...sequencesToAnimate].reverse()).reduce((previous, current, currentIndex) => {
        const { sequenceId, sequenceIndex, ...restAttributes } = current;
        const { duration, delay, overlay } = restAttributes;
        const id = getSequenceId(sequenceIndex, sequenceId, currentIndex);
        const totalDuration = calculateTotalDuration({ duration, delay, overlay }) + previous;

        localAnimationState[id] = {
          play,
          delay: currentIndex === 0 ? delay || 0 : previous,
        };

        return totalDuration;
      }, 0);

      setAnimationStates(localAnimationState);
    },
    [play],
  );

  // @ts-ignore
  return (
    <AnimateContext.Provider value={{ animationStates, register }}>
      {/* @ts-ignore */}
      {children}
    </AnimateContext.Provider>
  );
}
