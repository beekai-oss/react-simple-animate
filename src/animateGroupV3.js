// @flow
import React, { type Element, useState, useRef, useEffect } from 'react';
import type { AnimationType, AnimationStateType } from './types';
import calculateTotalDuration from './utils/calculateTotalDuration';

type Sequence = AnimationType & {
  sequenceId?: string | number,
  sequenceIndex?: number,
};

type Sequences = Array<Sequence>;

type Props = {
  play: boolean,
  sequences?: Sequences,
  reverseSequences?: Sequences,
  children: Element<*>,
};

type State = {
  animationStates?: AnimationStateType,
};

// $FlowIgnoreLine
export const AnimateContext = React.createContext({
  animationStates: {},
  register: () => {},
});

export default function AnimateGroup(props: Props) {
  const [animationStates, setAnimationStates] = useState();
  const animationsRef = useRef({});

  const register = props => {
    const { sequenceIndex, sequenceId } = props;
    const id = sequenceId || sequenceIndex;
    if (id === undefined || (sequenceIndex && sequenceIndex < 0) || (sequenceId && sequenceId === '')) return;

    animationsRef.current[id] = {
      ...props,
    };

    console.log(animationsRef.current);
  };

  useEffect(() => {
    const { sequences, play } = props;
    const sequencesToAnimate =
      Array.isArray(sequences) && sequences.length ? sequences : Object.values(animationsRef.current);
    const reverseSequencesToAnimate = [...sequencesToAnimate].reverse();

    return (play ? sequencesToAnimate : reverseSequencesToAnimate).reduce((previous, current, currentIndex) => {
      const { sequenceId, sequenceIndex, ...restAttributes } = current;
      const id = sequenceId === undefined && sequenceIndex === undefined ? currentIndex : sequenceId || sequenceIndex;
      const totalDuration =
        previous +
        calculateTotalDuration({
          ...animationsRef.current[id],
          play,
          restAttributes,
        });

      console.log('totalDuration', totalDuration);
      return totalDuration;
    }, 0);
  });

  return <AnimateContext.Provider value={{ animationStates, register }}>{props.children}</AnimateContext.Provider>;
}
