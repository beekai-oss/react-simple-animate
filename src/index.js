// @flow
import Animate from './animate';
import AnimateGroup from './animateGroup';

export { Animate, AnimateGroup };
