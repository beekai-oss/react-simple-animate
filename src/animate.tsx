import * as React from 'react';
import { AnimateContext } from './animateGroup';
import secToMs from './utils/secToMs';
import { AnimationProps } from './types';
import getSequenceId from "./utils/getSequenceId";

const { useEffect, useState, useRef, useContext } = React;

export default function Animate(props: AnimationProps) {
  const {
    play,
    children,
    render,
    start,
    end,
    complete = '',
    onComplete,
    delay = 0,
    duration = 0.3,
    easeType = 'linear',
    sequenceId,
    sequenceIndex,
  } = props;
  const onCompleteTimeRef = useRef(null);
  const [style, setStyle] = useState(start || {});
  const { register, animationStates = {} } = useContext(AnimateContext);
  const id = getSequenceId(sequenceIndex, sequenceId);

  useEffect((): void => {
    if ((sequenceIndex !== undefined && sequenceIndex >= 0) || sequenceId) register(props);
  }, []);

  useEffect(
    (): any => {
      setStyle({
        ...(play || (animationStates[id] || {}).play ? end : start),
        transition: `all ${duration}s ${easeType} ${parseFloat((animationStates[id] || {}).delay || delay)}s`,
      });

      if (play && (complete || onComplete)) {
        // @ts-ignore
        onCompleteTimeRef.current = setTimeout((): void => {
          complete && setStyle(complete);
          onComplete && onComplete();
        }, secToMs(parseFloat((animationStates[id] || {}).delay || delay) + duration));
      }

      return (): void => {
        // @ts-ignore
        onCompleteTimeRef.current && clearTimeout(onCompleteTimeRef.current);
      };
    },
    [id, animationStates, play, duration, easeType, delay, onComplete, start, end, complete],
  );

  return render ? render({ style }) : <div style={style}>{children}</div>;
}
