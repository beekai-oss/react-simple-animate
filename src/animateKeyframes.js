// @flow
import React from 'react';
import createStyle from './style/createStyle';
import createRandomName from './utils/createRandomName';
import { AnimateContext } from './animateGroup';

type Props = {
  keyframes: Array<string>,
  easeType?: string,
  durationSeconds?: number,
  render?: Object => any,
  play: boolean,
  playState?: boolean,
  delaySeconds?: number,
  direction?: string,
  iterations?: string | number,
  fillMode?: string,
  iterationCount?: number,
};

type State = {};

const styleTagName = 'style[data-id-rsi]';

export class AnimateKeyframes extends React.PureComponent<Props, State> {
  static defaultProps = {
    durationSeconds: 0.3,
    delaySeconds: 0,
    easeType: 'linear',
    render: undefined,
    iterations: 1,
    playState: 'running',
    direction: 'normal',
    fillMode: 'none',
    iterationCount: 1,
  };

  componentDidMount() {
    const { keyframes, durationSeconds, easeType } = this.props;

    this.animationName = createRandomName();
    this.styleTag = document.querySelector(styleTagName);
    if (!this.styleTag) {
      this.styleTag = document.createElement('style');
      this.styleTag.setAttribute('data-id', 'rsi');
      document.head.appendChild(this.styleTag);
    }
    this.index = this.styleTag.sheet.cssRules.length;

    this.styleTag.sheet.insertRule(
      createStyle({
        keyframes,
        durationSeconds,
        easeType,
        animationName: this.animationName,
      }),
      this.index,
    );
  }

  static getDerivedStateFromProps(nextProps: Props, prevState: State) {
    const { animationStates, play, sequenceId, sequenceIndex } = nextProps;
    const id = sequenceId || sequenceIndex;
    let currentPlay = play;

    if (id && animationStates && animationStates[id]) {
      const state = animationStates[id];
      currentPlay = state.play;
    }

    return {
      ...(currentPlay !== prevState.play ? { play: currentPlay } : null),
    };
  }

  componentWillUnmount() {
    this.styleTag.sheet.deleteRule(this.index);
  }

  animationName: string;
  index: number;
  styleTag: {};

  render() {
    const {
      children,
      play,
      render,
      playState,
      delaySeconds,
      direction,
      durationSeconds,
      easeType,
      iterations,
      fillMode,
    } = this.props;
    const style = play
      ? {
          animation: `${durationSeconds}s ${easeType} ${delaySeconds}s ${iterations} ${direction} ${fillMode} ${playState} ${
            this.animationName
          }`,
        }
      : null;

    return render ? render(style) : <div style={style}>{children}</div>;
  }
}

// $FlowIgnoreLine: flow complain about React.forwardRef disable for now
export default React.forwardRef((props: Props, ref) => (
  <AnimateContext.Consumer>
    {({ animationStates = {}, register = undefined }) => (
      <AnimateKeyframes {...{ ...props, animationStates, register }} forwardedRef={ref} />
    )}
  </AnimateContext.Consumer>
));
