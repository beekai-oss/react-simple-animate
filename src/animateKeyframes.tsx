import * as React from 'react';
import createTag from './logic/createTag';
import createRandomName from './utils/createRandomName';
import deleteRule from './logic/deleteRules';
import { AnimateContext } from './animateGroup';
import { AnimateKeyframesProps } from './types';

const { useRef, useEffect, useContext, useState } = React;

export default function AnimateKeyframes(props: AnimateKeyframesProps) {
  const {
    children,
    play,
    render,
    duration = 0.3,
    delay = 0,
    easeType = 'linear',
    playState = 'running',
    direction = 'normal',
    fillMode = 'none',
    iterationCount = 1,
    keyframes,
  } = props;
  const animationNameRef = useRef('');
  const styleTagRef = useRef({
    sheet: {},
  });
  const { register } = useContext(AnimateContext);
  const forceUpdate = useState(false)[1];

  useEffect(() => {
    animationNameRef.current = createRandomName();
    const { styleTag } = createTag({
      animationName: animationNameRef.current,
      keyframes,
    });
    styleTagRef.current = styleTag;
    register(props);

    if (play) forceUpdate(true);
    return (): void => deleteRule(styleTagRef.current.sheet, animationNameRef.current);
  }, []);

  const style = play
    ? {
        animation: `${duration}s ${easeType} ${delay}s ${iterationCount} ${direction} ${fillMode} ${playState} ${animationNameRef.current ||
          ''}`,
      }
    : null;

  return render ? render({ style }) : <div style={style || {}}>{children}</div>;
}
