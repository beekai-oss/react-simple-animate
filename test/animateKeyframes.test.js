import renderer from 'react-test-renderer';
import React from 'react';
import { AnimateKeyframesChild } from '../src/animateKeyframes';

jest.mock('../src/style/createTag', () => () => 'createTag');

describe('AnimateKeyframesChild', () => {
  const props = {
    keyframes: [],
    play: false,
    animationStates: [],
  };

  it('should render correctly', () => {
    const tree = renderer.create(<AnimateKeyframesChild {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when play set true', () => {
    const tree = renderer.create(<AnimateKeyframesChild {...{ ...props, play: true }} />);
    expect(tree).toMatchSnapshot();
  });
});
