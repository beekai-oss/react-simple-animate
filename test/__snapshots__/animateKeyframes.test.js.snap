// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimateKeyframesChild should render correctly 1`] = `<div />`;

exports[`AnimateKeyframesChild should render correctly when play set true 1`] = `
<div
  style={
    Object {
      "animation": "0.3s linear 0s 1 normal none running undefined",
    }
  }
/>
`;
